sdk/python/lib/pulumi/output.py

```diff
@@ -224,9 +224,12 @@ async def run() -> U:
                     # Try and set the result. This might fail if we're shutting down,
                     # so swallow that error if that occurs.
                     try:
-                        result_resources.set_result(resources)
-                        result_is_known.set_result(False)
-                        result_is_secret.set_result(False)
+                        if not result_resources.done():
+                            result_resources.set_result(resources)
+                        if not result_is_known.done():
```

Comment:
Check on L223 `if not result_is_known.done():` is perhaps redundant here now.

Reply:
Perhaps not because this is all async. And on that all of these checks are probably unsafe and should be written in EAFP style instead:
```
try:
   result_resources.set_result(resources)
except asyncio.InvalidStateError:
   pass
```

Reply:
Ah in case a concurrent thread would do a set_result in between the lines? I think that shouldn't happen as these futures are owned by the `def` so there's no concurrent code that sets them.

Reply:
I think I'd still rather EAFP style, more pythonic

Reply:
Irrespective of whether you take the EAFP or LBYL approach, there shouldn't be any race conditions since only one task is ever executing (in the Python interpreter) at a time, and the synchronisation points are explicitly specified where you `await`. I agree the check on L223 is redundant since there is no `await` between that check and this new one.

Regarding EAFP, I usually advocate for [contextlib.suppress](https://docs.python.org/3/library/contextlib.html#contextlib.suppress):

```python
with contextlib.suppress(asyncio.InvalidStateError):
    result_resources.set_result(resources)
with contextlib.suppress(asyncio.InvalidStateError):
    result_is_known.set_result(False)
with contextlib.suppress(asyncio.InvalidStateError):
    result_is_secret.set_result(False)
```